cmd/zilstat.in

```diff
@@ -0,0 +1,470 @@
+				invalid.append(ele)
+			elif ele not in cols:
+				invalid.append(ele)
+
+		if len(invalid) > 0:
+			sys.stderr.write("Invalid column definition! -- %s\n" % invalid)
+			usage()
+
+	if pool_name and dataset_name:
+		print ("Error: Can not filter both dataset and pool")
+		sys.exit(0)
+
+if sys.platform.startswith('freebsd'):
+	# Requires py-sysctl on FreeBSD
+	import sysctl
+
+	def kstat_update(pool = None, objid = None):
+		global kstat
+		try:
+			kstat = {}
+			if pool is None:
+				file = "kstat.zfs.misc.zil"
+				k = [ctl for ctl in sysctl.filter(file) \
+				    if ctl.type != sysctl.CTLTYPE_NODE]
+				kstat_process_str(k, "GLOBAL", len(file + "."))
+			elif objid:
+				file = "kstat.zfs." + pool + ".dataset.objset-" + objid
+				k = [ctl for ctl in sysctl.filter(file) if ctl.type \
+				    != sysctl.CTLTYPE_NODE]
+				kstat_process_str(k, objid, len(file + "."))
+			else:
+				file = "kstat.zfs." + pool + ".dataset"
+				zil_start = len(file + ".")
+				obj_start = len("kstat.zfs." + pool + ".")
+				k = [ctl for ctl in sysctl.filter(file)
+					if ctl.type != sysctl.CTLTYPE_NODE]
+				for s in k:
+					if not s or (s.name.find("zil") == -1 and \
+					    s.name.find("dataset_name") == -1):
+						continue
+					name, value = s.name, s.value
+					objid = re.findall(r'0x[0-9A-F]+', \
+					    name[obj_start:], re.I)[0]
+					if objid not in kstat:
+						kstat[objid] = dict()
+					zil_start = len(file + ".objset-" + \
+					    objid + ".")
+					kstat[objid][name[zil_start:]] = value \
+					    if (name.find("dataset_name")) \
+					    else int(value)
+
+		except IOError as e:
+			if e.errno == errno.EPIPE:
+				sys.exit(1)
+
+	def kstat_process_str(k, objset = "GLOBAL", zil_start = 0):
+			global kstat
+			if not k:
+				sys.exit(1)
+			kstat[objset] = dict()
+			for s in k:
+				if not s or (s.name.find("zil") == -1 and \
+				    s.name.find("dataset_name") == -1):
+					continue
+				name, value = s.name, s.value
+				kstat[objset][name[zil_start:]] = value \
+				    if (name.find("dataset_name")) else int(value)
+
+elif sys.platform.startswith('linux'):
+	def kstat_update(pool = None, objid = None):
+		global kstat
+		try:
+			kstat = {}
+			if pool is None:
+				k = [line.strip() for line in \
+				    open("/proc/spl/kstat/zfs/zil")]
+				kstat_process_str(k)
+			elif objid:
+				file = "/proc/spl/kstat/zfs/" + pool + "/objset-" + objid
+				k = [line.strip() for line in open(file)]
+				kstat_process_str(k, objid)
+			else:
+				objsets = os.listdir(f'/proc/spl/kstat/zfs/{pool}')
+				for objid in objsets:
+					if objid.find("objset-") == -1:
+						continue
+					file = "/proc/spl/kstat/zfs/" + pool + "/" + objid
+					k = [line.strip() for line in open(file)]
+					kstat_process_str(k, objid.replace("objset-", ""))
+		except IOError as e:
+			if e.errno == errno.EPIPE:
+				sys.exit(1)
+
+	def kstat_process_str(k, objset = "GLOBAL", zil_start = 0):
+			global kstat
+			if not k:
+				sys.exit(1)
+
```

Comment:
if this occurs, the user has no clue as to what happened. Surely we can be more user-friendly?